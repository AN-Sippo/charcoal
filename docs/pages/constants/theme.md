# `@charcoal/theme`

charcoal 全体で制約されるテーマ構造の型を提供します。

## インストール

npm

```bash
npm i @charcoal/theme
```

yarn

```bash
yarn add @charcoal/theme
```
