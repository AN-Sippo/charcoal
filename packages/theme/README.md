# `@charcoal/theme`

charcoal 全体で制約されるテーマ構造の型を提供します。

## インストール

npm

```
npm i --save-dev @charcoal/theme
```

yarn

```
yarn add -D @charcoal/theme
```
