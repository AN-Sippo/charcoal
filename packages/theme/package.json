{
  "name": "@charcoal-ui/theme",
  "version": "4.0.0-beta.15",
  "license": "Apache-2.0",
  "main": "./dist/index.cjs.js",
  "module": "./dist/index.esm.js",
  "exports": {
    "types": "./dist/index.d.ts",
    "require": "./dist/index.cjs.js",
    "import": "./dist/index.esm.js",
    "default": "./dist/index.esm.js"
  },
  "types": "./dist/index.d.ts",
  "sideEffects": [
    "*.css"
  ],
  "scripts": {
    "build": "npm-run-all --print-label --parallel 'build:*' --sequential serialize",
    "build:bundle": "FORCE_COLOR=1 tsup",
    "build:dts": "tsc --project tsconfig.build.json --pretty --emitDeclarationOnly",
    "build:copy-css": "cpx 'src/css/**/*.css' dist/css && cpx 'src/json/**/*.json' dist/json",
    "serialize": "node cli/index.js",
    "typecheck": "run-p --print-label 'typecheck:*'",
    "typecheck:main": "tsc --project tsconfig.build.json --pretty --noEmit",
    "typecheck:cli": "tsc --project cli/tsconfig.build.json --noEmit",
    "clean": "rimraf dist .tsbuildinfo",
    "test": "vitest run --passWithNoTests"
  },
  "devDependencies": {
    "cpx": "^1.5.0",
    "npm-run-all": "^4.1.5",
    "rimraf": "^3.0.2",
    "tsup": "^6.5.0",
    "typescript": "^4.9.5",
    "vitest": "^2.0.2"
  },
  "dependencies": {
    "@charcoal-ui/foundation": "^4.0.0-beta.15",
    "@charcoal-ui/utils": "^4.0.0-beta.15",
    "polished": "^4.1.4"
  },
  "files": [
    "src",
    "dist",
    "theme.json"
  ],
  "publishConfig": {
    "access": "public"
  },
  "repository": {
    "type": "git",
    "url": "https://github.com/pixiv/charcoal.git",
    "directory": "packages/theme"
  },
  "gitHead": "e1ece2e43901ae667afdd5c178040607d939dcd5"
}
