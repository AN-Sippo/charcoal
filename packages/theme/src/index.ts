import { Effect, Effects, OpacityEffect } from './effect'

export default {}

export * from './effect'

export type Material = string
export type GradientMaterial = { color: Material; ratio: number }[]

/**
 * @deprecated
 */
// eslint-disable-next-line @typescript-eslint/consistent-type-definitions
type ThemeMaterials = {
  background1: Material
  background2: Material
  surface1: Material
  surface2: Material
  /**
   * Black Fade 20にするとボーダーと共通化できる
   */
  surface3: Material
  surface4: Material
  // NOTE: the current definition of surface5 is invalid
  surface6: Material
  /**
   * サムネイルに被せる色。
   * Lightモードの色が2と共通。コントラストの低下を狙うなら2でいいが、サムネイルで色味を調整しすぎる感がある。
   */
  surface7: Material
  /**
   * 磨りガラス効果を使えれば6と統合できそう
   */
  surface8: Material
  surface9: Material
  link1: Material
  link2: Material
  text1: Material
  /**
   * 1と2は統合できるかもしれない
   */
  text2: Material
  text3: Material
  /**
   * Surface 4に重ねた時に薄すぎる感あり
   */
  text4: Material
  text5: Material
  icon6: Material
  border: Material
}

/**
 * @deprecated
 */
export interface ThemeEffects {
  hover: Effects
  press: Effects
  disabled: Effects
}

// eslint-disable-next-line @typescript-eslint/consistent-type-definitions
export type ThemeEffect = {
  hover: Effect
  press: Effect
}

// eslint-disable-next-line @typescript-eslint/consistent-type-definitions
export type ThemeElementEffect = {
  disabled: OpacityEffect
}

/**
 * @deprecated
 */
// eslint-disable-next-line @typescript-eslint/consistent-type-definitions
type ThemeSemantic = {
  brand: Material
}

/**
 * @deprecated Use `ThemeColor` instead
 */
export type ThemeColors = ThemeMaterials &
  ThemeSemantic & {
    premium: Material
  }

// eslint-disable-next-line @typescript-eslint/consistent-type-definitions
export type ThemeColor = {
  transparent: Material
  background1: Material
  background2: Material
  surface1: Material
  surface2: Material
  surface3: Material
  surface4: Material
  surface6: Material
  surface7: Material
  surface8: Material
  surface9: Material
  link1: Material
  link2: Material
  text1: Material
  text2: Material
  text3: Material
  text4: Material
  text5: Material
  icon6: Material
  border: Material
  brand: Material
  assertive: Material
  /**
   * @deprecated
   */
  premium: Material
}

// eslint-disable-next-line @typescript-eslint/consistent-type-definitions
export type ThemeColorGradient = {
  surface4Gradient: GradientMaterial
}

export interface Theme {
  /**
   * @deprecated
   *
   * Use `color` instead.
   */
  materials: ThemeMaterials
  /**
   * @deprecated
   *
   * Use `color` instead.
   */
  semantic: ThemeSemantic
  /**
   * @deprecated Use `color` instead
   */
  colors: ThemeColors
  color: ThemeColor
  gradientColor: ThemeColorGradient
  /**
   * @deprecated Use `effect` instead
   */
  effects: ThemeEffects
  /**
   * effect for color (background or font color)
   */
  effect: ThemeEffect
  /**
   * effect for element its own (opacify element includes descendants)
   */
  elementEffect: ThemeElementEffect
}
