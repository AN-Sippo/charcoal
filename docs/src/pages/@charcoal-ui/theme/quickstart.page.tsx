import Link from 'next/link'
import { ContentRoot } from '../../../components/ContentRoot'

import { InlineCode } from '../../../components/InlineCode'
import { SSRHighlight } from '../../../components/SSRHighlight'

export default function InstallPage() {
  return (
    <ContentRoot>
      <h1>@charcoal-ui/theme クイックスタート</h1>
      <Link href="https://www.npmjs.com/package/@charcoal-ui/theme">
        <img
          alt="npm-badge"
          src="https://img.shields.io/npm/v/@charcoal-ui/theme?label=%40charcoal-ui%2Ftheme&style=flat-square&logo=npm"
        />
      </Link>
      <h2>インストール</h2>
      <h3>npm</h3>
      <SSRHighlight
        code="npm install --save-dev @charcoal-ui/theme"
        lang="shell"
      />
      <h3>yarn</h3>
      <SSRHighlight code="yarn add -D @charcoal-ui/theme" lang="shell" />

      <hr />

      <h2>概要</h2>
      <p>
        <InlineCode>@charcoal-ui/theme</InlineCode>
        は、全体で使われるテーマ型と、デフォルトのライト・ダークテーマを提供します。
      </p>

      <h2>使い方</h2>
      <p>
        <InlineCode>styled-components</InlineCode>と合わせて使用できます。
      </p>
      <p>
        デフォルトのテーマとして提供している<InlineCode>light</InlineCode>、
        <InlineCode>dark</InlineCode>をそのまま使うこともできます。
      </p>
      <SSRHighlight code={sampleCode} lang="typescript" />
    </ContentRoot>
  )
}

const sampleCode = `
  import { light, dark } from '@charcoal-ui/theme'
  import { ThemeProvider } from 'styled-components'
  import { FC, ReactNode } from 'react'

  const LightExample: FC<{ children: ReactNode }> = ({ children }) => (
    <ThemeProvider theme={light}>
      {children}
    </ThemeProvider>
  )

  const DarkExample: FC<{ children: ReactNode }> = ({ children }) => (
    <ThemeProvider theme={dark}>
      {children}
    </ThemeProvider>
  )
`
