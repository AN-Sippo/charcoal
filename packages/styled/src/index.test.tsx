import { light } from '@charcoal-ui/theme'
import 'jest-styled-components'

import renderer from 'react-test-renderer'
import { ThemeProvider } from 'styled-components'
import { Example } from './index.story'
import { myTheme } from './storyHelper'

function render(children: JSX.Element) {
  return renderer
    .create(<ThemeProvider theme={myTheme(light)}>{children}</ThemeProvider>)
    .toJSON()
}

// TODO: もう少し theme() 関数に対する丁寧なユニットテストが欲しい
describe('Story', () => {
  test('<Example />', () => {
    expect(render(<Example />)).toMatchSnapshot()
  })
})
