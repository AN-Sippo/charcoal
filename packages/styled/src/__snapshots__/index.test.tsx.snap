// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Story <Example /> 1`] = `
<div
  className=""
>
  <div
    className=""
  >
    <div
      className=""
    >
      \${({ theme }) =&gt; theme.utils.typography(14, true)}
    </div>
  </div>
  <div
    className=""
  >
    <div
      className=""
    >
      <pre
        className=""
      >
        <code
          className=""
        >
          \${({ theme }) =&gt; [
    theme.utils.margin(16, 'auto'),
    theme.utils.padding(4, 8, 16, 24),
  ]}
        </code>
      </pre>
    </div>
  </div>
  <div
    className=""
  >
    <div
      className=""
    >
      theme.utils.assertiveRingCss
    </div>
  </div>
  <div
    className=""
  >
    <button
      className=""
    >
      theme.utils.focusVisibleFocusRingCss
    </button>
    <button
      className=""
      disabled={true}
    >
      theme.utils.disabledCss
    </button>
  </div>
</div>
`;

exports[`ThemeUtilsTest <ThemeUtilsTest /> 1`] = `
<div
  className=""
>
  <div
    className=""
  />
  <div
    className=""
  />
  <div
    className=""
  />
</div>
`;
